src/altinn-app-frontend/src/components/custom/CustomWebComponent.tsx

```diff
@@ -0,0 +1,94 @@
+import { getTextResourceByKey } from 'altinn-shared/utils';
+import * as React from 'react';
+import { useAppSelector } from 'src/common/hooks';
+import type { ITextResource, ITextResourceBindings } from 'src/types';
+import type { IComponentProps } from '..';
+
+export interface ICustomComponentProps extends IComponentProps {
+  tagName: string;
+}
+
+function CustomWebComponent({
+  tagName,
+  formData,
+  componentValidations,
+  textResourceBindings,
+  dataModelBindings,
+  language,
+  handleDataChange,
+  ...passThroughProps
+}: ICustomComponentProps) {
+  const Tag = tagName as any;
+  const wcRef = React.useRef(null);
+  const textResources = useAppSelector(
+    (state) => state.textResources.resources,
+  );
+
+  React.useLayoutEffect(() => {
+    const handleChange = (customEvent: CustomEvent) => {
+      const { value, field } = customEvent.detail;
+      handleDataChange(value, field);
+    };
+    const { current } = wcRef;
+    current.addEventListener('dataChanged', handleChange);
+    return () => {
+      current.removeEventListener('dataChanged', handleChange);
+    };
+  }, [handleDataChange, wcRef]);
+
+  React.useLayoutEffect(() => {
+    const { current } = wcRef;
+    current.texts = getTextsForComponent(
+      textResourceBindings,
+      textResources,
+      false,
+    );
+    current.dataModelBindings = dataModelBindings;
+    current.language = language;
+  }, [wcRef, textResourceBindings, textResources, dataModelBindings, language]);
+
+  React.useLayoutEffect(() => {
+    const { current } = wcRef;
+    current.formData = formData;
+    current.componentValidations = componentValidations;
+  }, [formData, componentValidations]);
+
+  if (!Tag || !textResources) {
+    return null;
+  }
+
+  const propsAsAttributes: any = {};
+  Object.keys(passThroughProps).forEach((key) => {
+    propsAsAttributes[key] = JSON.stringify(passThroughProps[key]);
```

Comment:
Won't this mean that simple string values from the form layout gets wrapped in double-quotes? Might get a bit annoying on the receiving end. I get that we can't pass complex objects as HTML attributes, but maybe we should filter those out (or assign them to the ref like with dataModelBindings etc above). Simple values like string/number/boolean could be passed through without JSON encoding.

Reply:
Hmm yep, you're right. Didn't think of that. Will do a filtering for more complex objects/arrays, and leave the rest as-is.